### ISSUE TITLE
Proposal: World Action Model (1 Hz camera multi-view encoder without BEV view  + encoded features temporal fusion)

### ISSUE THREAD

**Issue by intisar1020** [autowarefoundation/auto_e2e#93]
## Summary
This proposes a concrete, lightweight design for the World Action Model module, following the diagram's flow and discussion last meeting (Jun. 24) for the **World Action Model (WAM)** band:
following is from the latest design sketch by @m-zain-khawaja.
```
multi-camera images (7) @ 1Hz --> Vision Backbone (reuse, FROZEN) --> Feature Embedding
   --> Encoded Visual History --> Feature Decoding --> Visual Future Feature Prediction
   --> Future Feature Reconstruction Loss (JEPA)
```

My understanding of the update is that now the WAM's **Encoded Visual History** vector replaces the old BEV-derived `visual_history` (which is currently a zeros placeholder) and feeds two consumers: the **Reactive planner** and (later) the **Reasoning model**. The reconstruction loss is in **multi-view feature space, NOT BEV**.

This is a design sketch for discussion. Corrections and advice are very welcome.

## Design principles
- **Cheap.** The WAM runs at 1 Hz and only adds *small* trainable heads on top of a frozen backbone. Combine/compress/decode all operate on tiny token sets.
- **Reuse.** Lean on existing repo pieces: `backbone.py`, the per-view pooling in `FeatureFusion`, will copy the cross-attention pattern from `view_fusion`/`bev_fusion`, the GRU in `temporal_memory/one_hz_encoder.py` (or proposed a temporal attention-pool), and the `feature_reconstruction_loss.py` comparator.
- **Decoupled from the reactive trunk** (see backbone freezing below).

---

## Component 1  Vision Backbone (reuse + **FREEZE**)

Reuse the existing backbone (same arch/pretrained weights):
```python
x = camera_tiles.reshape(B * V, C, H, W)
features = self.Backbone(x)          # reuse but with requires_grad=False
```
**Freeze it.** Rationale (three wins):
1. **No conflicting gradients.** The backbone is also trained by `Reactive_E2E` for BEV + trajectory. If the WAM's JEPA loss also updated the backbone, the two objectives might fight (BEV/trajectory-optimal features vs. future-prediction-optimal features). Freezing removes that conflict. But we can also use a dedicated backbone copy for WAM (frozen) and Reactive (trainable) if we want to keep the two objectives separate.
2. **Stationary JEPA target (anti-collapse).** A frozen backbone gives a **fixed** target feature space, so the predictor can't trivially collapse (the BYOL/JEPA requirement for a non-trainable target encoder). The **same frozen backbone** encodes both the past (context) and the future (targets), no separate frozen copy / EMA needed.
3. **Cheap.** No backbone backprop; only the small WAM heads train.

Decision to confirm: **separate frozen copy (pretrained init)** vs **shared-with-stop-gradient** on Reactive's live backbone. Frozen pretrained copy is cleanest (stable targets, matches @m-zain-khawaja's frozen-init decision in #13); shared-stop-gradient is lighter but gives non-stationary targets as Reactive trains the backbone.

## Component 2  Feature Embedding (cheap cross-attention, BEV-style)

Per-frame, combine the V camera feature maps into **one per-frame embedding vector** via a **single-query cross-attention**. Thinking to adopt the `bev_fusion` learnable-query idea collapsed to one non-geometric query (no projection, since we're not in BEV), with learned camera embeddings borrowed from `CrossAttentionViewFusion`:

```python
class ViewAttentionPool(nn.Module):          # cheaper option: 1 layer, few heads
    def __init__(self, embed_dim=256, num_views=7, num_heads=4):
        super().__init__()
        self.view_embed = nn.Parameter(torch.randn(1, num_views, embed_dim))
        self.query      = nn.Parameter(torch.randn(1, 1, embed_dim))
        self.attn       = nn.MultiheadAttention(embed_dim, num_heads, batch_first=True)
        self.norm       = nn.LayerNorm(embed_dim)

    def forward(self, per_view_feat, B, V):   # per_view_feat: [B*V, C, h, w]
        C   = per_view_feat.shape[1]
        tok = per_view_feat.mean(dim=(2, 3)).reshape(B, V, C)  # pool each view --> [B,V,C]  
        tok = tok + self.view_embed                            # camera identity or learned positional embedding for the cameras
        q   = self.query.expand(B, -1, -1)
        out, _ = self.attn(q, tok, tok)                        # attend over V = 7 keys (asuming 7 cam)
        return self.norm(out.squeeze(1))                       # [B, C] per-frame embedding
```
Cost is negligible (attention over 7 tokens). Reuse `FeatureFusion`'s per-view multi-scale pool + `channel_proj` (1440-->256) to produce `per_view_feat`.

## Component 3  Encoded Visual History (compress the N-frame stack, cheap)

Maintain a **rolling stack of the last N per-frame embeddings** `[B, N, D]` (at 1 Hz). Compress to a single **Encoded Visual History** vector. Two options:

- **A. Temporal attention-pool (preferred --> consistent with Component 2).** A single learned query + **temporal positional embeddings** cross-attends over the N frame tokens --> `[B, D]`. This is the *same attention-pool pattern* as the view combiner, just applied over time axis instead of views, so the WAM uses one consistent primitive (cross-attention pooling) for both the spatial (multi-view) and temporal (multi-frame) aggregation. Cheap: 1 layer, few heads, attention over only N≈6 tokens.
  ```python
  tok = frame_embeddings + self.time_pos        # [B, N, D]  + temporal positional embeds
  q   = self.history_query.expand(B, -1, -1)    # [B, 1, D]
  out, _ = self.history_attn(q, tok, tok)       # attend over N≈6 frame tokens
  encoded_visual_history = self.norm(out.squeeze(1))   # [B, D]
  ```
- **B. GRU summary (reuse alternative).** Run a GRU over the N frame embeddings; final hidden = encoded history — the summarization half of `HistoryEncoder` in `temporal_memory/one_hz_encoder.py`. Marginally cheaper, but introduces a second aggregation primitive; kept as a fallback.

**Efficiency note:** because the backbone is frozen and the per-frame embedding depends only on its own frame, at 1 Hz you encode **only the new frame** each step and **cache** the previous N−1 embeddings. So per-step cost is one frame through the frozen backbone + the cheap compressor.

Output: `encoded_visual_history` `[B, D]` --> feeds the Reactive planner (replaces the zeros `visual_history`).

## Component 4 Feature Decoding (JEPA predictor, cheap & parallel)

Predict the **next N future per-frame multi-view features** (symmetric: `T_future = T_past = N`) from the encoded history. Use a **parallel query decoder** (no autoregression):

```python
# N learned future-step queries (+ temporal positional embeddings) cross-attend to the
# encoded history (or to the N past frame embeddings), then project to the frozen-backbone
# feature space (per-view pooled).
future_queries = self.future_pos.expand(B, -1, -1)            # [B, N, D]
dec, _ = self.decoder_attn(future_queries, context, context)  # 1-layer cross-attn
future_pred = self.to_feat(dec)                               # [B, N, V*C]  (per-view pooled features)
```
Cheap: one cross-attention layer + a linear projection. alternative: a GRU rollout for N steps (also cheap), but the parallel query decoder avoids sequential latency.

## Component 5  JEPA Loss

- **Targets:** the **frozen backbone** applied to the actual **N future frames**, per-view pooled --> `[B, N, V*C]`, **detached** (no grad). Because the backbone is frozen, targets are fixed -> no collapse, no EMA needed.
- **Loss:** MSE between `future_pred` and targets, optional temporal weighting (near-future weighted higher). **Reuse** `feature_reconstruction_loss.py`  this comparator (weighted per-step MSE) was **implemented by @gcordova10 in #80**, realising @riita10069's JEPA proposal (#13). We adopt it directly here — in practice, construct it with `num_future_steps=N` (its default is 4) and pass N per-step tensors (adapting the per-frame multi-view format).

```python
L_jepa = feature_reconstruction_loss(future_pred, target_future_feats.detach())
# Total: L = L_traj + λ * L_jepa
```

---

## I/O contract

```python
class WorldActionModel(nn.Module):
    def forward(self, multi_view_stream):    # [B, N, V, 3, H, W] @ 1Hz (past)
        ...
        return encoded_visual_history        # [B, D]  --> planner

    def compute_jepa_loss(self, multi_view_stream, future_frames):
        # future_frames: [B, N, V, 3, H, W] (frozen-backbone targets)
        return L_jepa, encoded_visual_history
```
Integration in `auto_e2e.py`: build `self.World_Action_Model`; in `forward`, run it on the 1 Hz stream --> `encoded_visual_history` --> pass into `Reactive_E2E` (replacing the input `visual_history`); surface `L_jepa` in train mode.

## Reuse map

| WAM piece | Reuse from |
|---|---|
| Vision backbone (frozen) | `backbone.py` |
| Per-view pool + channel_proj | `FeatureFusion` (pre-BEV path) |
| View cross-attention + camera embeddings | `view_fusion/bev_fusion.py` (learned query), `CrossAttentionViewFusion` (view embeds) |
| JEPA loss comparator | `losses/feature_reconstruction_loss.py` (implemented by @gcordova10, #80) |

## Key design decisions (to settle in the issue)

| Decision | Options | Lean |
|---|---|---|
| Backbone | separate frozen copy vs shared+stop-grad | frozen copy (stable targets) |
| Future-target granularity | pooled per-view vs full spatial maps | **pooled** (v1, cheap) |
| Temporal compressor | temporal attention-pool vs GRU (reuse) | **temporal attention-pool** (consistent with the view embedding) |
| Decoder | parallel query decoder vs GRU rollout | parallel (no latency) |
| `encoded_visual_history` dim | keep 896 (drop-in to planner) vs 256 (cleaner, needs planner `visual_history_dim` change) | TBD |
| N (= past = future frames) | e.g. 6 (6.4 s @ 1 Hz) | TBD |
| Action-conditioning | unconditioned vs condition future on planned trajectory (GAIA-style) | for now unconditioned for simplicity |

## Implementation plan (phased, additive)

1. **Module skeleton + synthetic test.** `WorldActionModel` with the I/O contract above, tested on random tensors (repo pattern), lands before data. Default off / zeros fallback so `Reactive_E2E` keeps working.
2. **Components 2-3** (embedding + compressor) --> produce a real `encoded_visual_history`.
3. **Components 4-5** (decoder + JEPA loss) with frozen-backbone targets.
4. **Wire into `auto_e2e.py`** behind a flag; surface `L_jepa` in train mode.
5. **Dataloader extension** (separate, ties to #16): emit the 1 Hz multi-view **past + future** stream `[B, N, V, 3, H, W]`.

## Open questions
- `encoded_visual_history` dim, keep 896 or move to 256 (planner change)?
- Pooled vs spatial future targets (cost vs richness)?
- Should the future prediction be **action-conditioned** (future views depend on where we drive)?
- N and the 1 Hz sampling, fixed or tunable per dataset?

@m-zain-khawaja I would be glad to contribute and start with component 2 and 3.

## References
- Architecture diagram (World Action Model @ 1 Hz band)
- #13 FutureState JEPA proposal (@riita10069); #80 feature-reconstruction loss + action conditioning (@gcordova10) — reused here; frozen-backbone target decision (@m-zain-khawaja, #13); #82 temporal memory / `visual_history` interface
- **Credits:** the JEPA reconstruction loss this proposal builds on was implemented by @gcordova10 (#80) from @riita10069's proposal (#13); the frozen-backbone target choice follows @m-zain-khawaja.
- JEPA (LeCun 2022); BYOL (Grill et al. 2020); MILE (Hu et al. 2022); GAIA-1 (Hu et al. 2023)

**Comment by gcordova10** [comment]
Thanks @intisar1020 — this matches very closely what I implemented right after @m-zain-khawaja's answers in the 24/06 meeting (his 5 interface answers + the miro diagram). It's now up as **PR #85** (`feat/t2-jepa-target-encoder`), with the `World_Action_Model_E2E` placeholder implemented **and wired into `AutoE2E`** — happy to take it forward together so we don't duplicate.

**How my implementation maps to your sketch:**
| Your flow | What's implemented (`world_action_model.py`) |
|---|---|
| Vision Backbone → Feature Embedding | `FrameEncoder`: shared backbone → GAP → linear → **224-d** per-frame embedding |
| Encoded Visual History | `RollingHistoryBuffer` (FIFO, len **4**) → concat → **896** (= `visual_history_dim`); replaces the zeros placeholder, fed to the reactive planner (and later Reasoning) |
| Feature Decoding → Visual Future Feature Prediction | `future_predictor` → N=4 future features |
| Future Feature Reconstruction Loss (JEPA) | `FeatureReconstructionLoss` (with the **L1** option) vs a **frozen copy** of the encoder (`JepaTargetEncoder`, stop-gradient) — multi-view feature space, not BEV |

It reuses the repo pieces you mention (`feature_reconstruction_loss.py`, a frozen target encoder), is additive + opt-in (`enable_world_model`, default off), and tests pass locally (mypy/ruff clean). CI is currently gated by #88.

**A couple of points to align (your input welcome):**
1. **Backbone freezing.** You wrote "reuse, FROZEN". In the meeting @m-zain-khawaja  said keep one shared backbone where the *online* copy is **trainable** and the **target** is a **frozen copy** (the standard JEPA anti-collapse setup) — which is what I built. Do we want the whole WAM backbone frozen (only small heads train), or trainable-online + frozen-target? Either works; I'd lean to Zain's.
2. **History temporal fusion.** I went with a simple concat-FIFO (4×224→896) to match the diagram's "Rolling FIFO Buffer (4)". Your **temporal attention-pool** (or the `one_hz_encoder` GRU) is a clean drop-in upgrade for that step — easy to swap behind the same interface.
3. **Prediction granularity.** I currently predict pooled per-frame embeddings (224, per the miro), but `TRIAL.md` and Zain's #85 comment point to **per-view feature maps** `[B, 1440, 7, 7]` × 4 as the JEPA target. I've asked Zain to confirm the level on #85 — the loss already handles feature-map lists, so it's a contained switch either way.

Want to co-author this on PR #85? Your temporal-attention history head and the backbone-freezing decision are clean drop-ins behind this interface — the `AutoE2E` wiring is already done, and the 1 Hz/10 Hz data feeding is the separate #16 (`Data Loader`) item. Just flagging it's already built so we pool effort.

**Comment by intisar1020** [comment]
Thanks @gcordova10, thats great, and no sense duplicating. Happy to co-author this on feat/world-model and build on your 79a2a54. Your mapping matches what I had in mind almost. 
just a quick alignments on your three points:
  
  1. Backbon. I 100% agreed, go with @m-zain-khawaja  setup. You r  right and that ones on me: I wrote "reuse, FROZEN" but conflated the target encoder with the whole backbone. Lets do the standard JEPA setup you built ->  trainable online encoder + frozen-copy target. The frozen copy gives the anti-collapse target, the online backbone still adapts, and the JEPA gradient on the shared backbone reads as a useful
  2. Temporal fusion, agreed, keep it swappable. Keep your concat-FIFO (4×224 --> 896) as the simple default, and based on the opinion of @m-zain-khawaja  I will proceed to  add a temporal attention-pool behind the same interface as an opt-in.  I think concat-FIFO is simple, straightforward with  the property that it  weights all 4 frames equally and is fixed-length and order is fixed. on the other hand, the attention-pool is a learnable aggregation. a single query attends over the N frame tokens (+ temporal positional embeddings), so the model (might)  weight frames by relevance (heopfully based on recency) and its length-flexible. One detail so its seemleslly drop-in. pooling over N tokens preserves the per-frame dim (224), so I will have to  add a small Linear(224 -> 896) after the  pool to match visual_history_dim, same interface, same 896 output. (The one_hz_encoder GRU is the other swappable option.)                                       

  3. Granularity, per-frame, agreed (for v1). Lets keep per-frame embeddings as the JEPA prediction/target for now. Since the loss already handles lists of feature maps, per-view [B,V,C,h,w] stays an easy later ablation if we want richer targets. but I think per frame is good and cheaper (224).

**Comment by gcordova10** [comment]
¡Gracias @intisar1020! Buenas noticias sobre el pool de atención temporal (tu punto 2): de hecho, ya tenía ese bloque: la `bev_queue` de #87 (autoatención temporal sobre el historial de fotogramas, desde el 23/06). Así que no hace falta crearlo dos veces; lo he integrado en el Modelo Mundial en #85 como un agregador **opcional** (`history_aggregator="attention"`, por defecto sigue siendo `concat-FIFO`), y he cerrado #87.

Sin embargo, esto es solo una parte de tu propuesta; hay muchas más en las que me encantaría contar con tu ayuda, por ejemplo, con tu `ViewAttentionPool` para la incrustación por fotograma (actualmente estoy usando un GAP+Linear simple en `FrameEncoder`). Estaré encantado de desarrollarlo juntos en #85.

**Comment by intisar1020** [comment]
Gracias @gcordova10 ! i am glad that the temporal pool was already covered by #87. I will add the ViewAttentionPool for the per-frame embedding as an opt-in option alongside the current view averaging in FrameEncoder  (view_aggregator="mean" or "attention", default stays "mean"). so if we enable the attention the model will do cross-attention over the cameras instead of equal weighting them. Jus to confrm, will I develop it in #85 or should I open a PR into your branch or add me to the fork?

**Comment by gcordova10** [comment]
@intisar1020 perfect — `view_aggregator="mean"(default)/"attention"` in `FrameEncoder` is exactly the right shape (opt-in, default unchanged).

On where to land it: let's keep **everything flowing to upstream `main`** (reviewed by @m-zain-khawaja) rather than PRs into my fork. The only reason it feels awkward right now is that `FrameEncoder` only lives on my #85 branch, not in `main` yet. So the clean sequence:

1. **#94 (CI fix) → #85 merge into `main`** → then `FrameEncoder` is in `main`.
2. **You open your `ViewAttentionPool` PR straight to `autowarefoundation/auto_e2e:main`** → @m-zain-khawaja reviews. (I'll do the same for the #13 JEPA-loss / `train_il` side.)

You can absolutely **start coding now** on a branch based off #85's code — just open the PR to `main` once #85 lands (rebase onto `main`) so it's a clean, focused diff for review.

Boundary so we don't overlap: **you** = the per-frame view embedding in `FrameEncoder` (#93); **me #13** = the JEPA loss in `train_il` + the prediction-error introspection (separate, only touches `training/`). Different files, no collision. Looking forward to it!
